File: window/optionswindow.py
from PyQt5.QtCore import QUrl
from PyQt5.QtGui import QDesktopServices
from window.template import OpaqueWindow


class OptionsWindow(OpaqueWindow):
  def __init__(self, engine, **kwargs):
    super().__init__(engine, 'qml/Options.qml', **kwargs)
    self.Global.optionsWindowOpenChanged.connect(self.onWindowOpenChanged)
    self.rootObject().openUrl.connect(self.openUrl)

  def onWindowOpenChanged(self):
    isOpen = self.Global.property('optionsWindowOpen')
    self.setVisible(isOpen)

  def openUrl(self, url):
    self.Global.setProperty('optionsWindowOpen', False)
    QDesktopServices.openUrl(QUrl(url))

File: common/src/window/optionswindow.cpp
#include "optionswindow.h"

OptionsWindow::OptionsWindow(QQmlEngine* engine, Settings* settings) : Window(engine, false, true)
{
  this->settings = settings;

  setSource(QUrl("qrc:/ui/Options.qml"));

  clientPathInput = rootObject()->findChild<QObject*>("clientPathInput");

  connect(global(), SIGNAL(optionsWindowOpenChanged()),
          this, SLOT(onWindowOpenChanged()));
  connect(rootObject(), SIGNAL(openUrl(QString)),
          this, SLOT(onOpenUrl(QString)));
  connect(rootObject(), SIGNAL(browseClientPath()),
          this, SLOT(onBrowseClientPath()));
  connect(rootObject(), SIGNAL(save()),
          this, SLOT(save()));
}

void OptionsWindow::onWindowOpenChanged()
{
  bool open = global()->property("optionsWindowOpen").toBool();

  if (open)
    load();

  setVisible(open);
}

void OptionsWindow::onBrowseClientPath()
{
  auto file = QFileDialog::getOpenFileName(this, "Find Game Client", "", "Path of Exile Client (Client.exe;PathOfExileSteam.exe)");
  if (!file.isEmpty())
     clientPathInput->setProperty("text", QFileInfo(file).dir().absolutePath());
}

void OptionsWindow::onOpenUrl(const QString& url)
{
  global()->setProperty("optionsWindowOpen", false);
  QDesktopServices::openUrl(QUrl(url));
}

void OptionsWindow::load()
{
  clientPathInput->setProperty("text", settings->value("poeClientPath").toString());
}

void OptionsWindow::save()
{
  settings->setValue("poeClientPath", clientPathInput->property("text"));
}
